File: foundational_design_patterns/6_planning/src/planning_plan_and_act.py
"""
Planning Pattern: Plan-and-Act with explicit replanning (LangGraph)

Anchor scenario: incident response automation.

Flow:
- planner node creates ordered actions
- executor node performs one action at a time
- reviewer node decides whether replanning is required
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass, field
from typing import Literal

from pathlib import Path

ROOT_DIR = next(parent for parent in Path(__file__).resolve().parents if (parent / "ssl_fix.py").exists())
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from repo_support import configure_example, get_default_model

configure_example(__file__)

from langchain_openai import ChatOpenAI
from langgraph.graph import END, StateGraph


@dataclass
class PlanActState:
    """State container for plan-and-act execution."""

    incident: str
    context: str
    plan: list[str] = field(default_factory=list)
    completed: list[str] = field(default_factory=list)
    current_action: str = ""
    notes: list[str] = field(default_factory=list)
    requires_replan: bool = False


llm = ChatOpenAI(model=get_default_model(), temperature=0)


def planner_node(state: PlanActState) -> PlanActState:
    """Create or refresh the execution plan."""
    prompt = f"""
You are an incident commander. Build a short action plan (4-6 steps) for this incident.

Incident:
{state.incident}

Context:
{state.context}

Already completed:
{state.completed}

Return one step per line, no numbering.
"""
    response = llm.invoke(prompt).content.strip().splitlines()
    plan = [line.strip("-• ").strip() for line in response if line.strip()]

    state.plan = [step for step in plan if step not in state.completed]
    state.notes.append(f"planner: generated {len(state.plan)} remaining steps")
    state.requires_replan = False
    return state


def executor_node(state: PlanActState) -> PlanActState:
    """Execute one action from the plan."""
    if not state.plan:
        state.current_action = ""
        return state

    action = state.plan.pop(0)
    state.current_action = action

    exec_prompt = f"""
You are executing one incident response action.

Action: {action}
Incident: {state.incident}

Return exactly two lines:
STATUS: done|blocked
NOTE: <short operational note>
"""
    result = llm.invoke(exec_prompt).content
    lines = [line.strip() for line in result.splitlines() if line.strip()]

    status = "done"
    note = "no note"
    for line in lines:
        if line.lower().startswith("status:"):
            status = line.split(":", 1)[1].strip().lower()
        if line.lower().startswith("note:"):
            note = line.split(":", 1)[1].strip()

    if status == "done":
        state.completed.append(action)
        state.notes.append(f"executor: done -> {action} | {note}")
        state.requires_replan = False
    else:
        state.notes.append(f"executor: blocked -> {action} | {note}")
        state.requires_replan = True

    return state


def reviewer_node(state: PlanActState) -> PlanActState:
    """Decide if another step can run or if replanning is needed."""
    if state.requires_replan:
        state.notes.append("reviewer: requesting replanning")
    elif state.plan:
        state.notes.append("reviewer: continue with remaining plan")
    else:
        state.notes.append("reviewer: plan complete")
    return state


def route_after_review(state: PlanActState) -> Literal["planner", "executor", "end"]:
    if state.requires_replan:
        return "planner"
    if state.plan:
        return "executor"
    return "end"


def build_graph():
    graph = StateGraph(PlanActState)
    graph.add_node("planner", planner_node)
    graph.add_node("executor", executor_node)
    graph.add_node("reviewer", reviewer_node)

    graph.set_entry_point("planner")
    graph.add_edge("planner", "executor")
    graph.add_edge("executor", "reviewer")
    graph.add_conditional_edges(
        "reviewer",
        route_after_review,
        {"planner": "planner", "executor": "executor", "end": END},
    )

    return graph.compile()


def main() -> None:
    print("\n" + "=" * 80)
    print("PLANNING PATTERN — PLAN-AND-ACT (LANGGRAPH)")
    print("=" * 80)

    if not os.getenv("OPENAI_API_KEY"):
        print("❌ OPENAI_API_KEY is required for this example.")
        return

    initial = PlanActState(
        incident="Payment API error rate jumped to 35% after deployment.",
        context="Service: billing-gateway, Region: eu-west, Recent change: release v2.8.4",
    )

    app = build_graph()
    final = app.invoke(initial)

    if isinstance(final, dict):
        completed = final.get("completed", [])
        notes = final.get("notes", [])
    else:
        completed = getattr(final, "completed", [])
        notes = getattr(final, "notes", [])

    print("\n✅ Completed steps:")
    for idx, step in enumerate(completed, 1):
        print(f"  {idx}. {step}")

    print("\n🧾 Execution notes:")
    for note in notes:
        print(f"  - {note}")


if __name__ == "__main__":
    main()

File: foundational_design_patterns/6_planning/typescript/src/planning_plan_and_act.ts
/**
 * Planning Pattern: Plan-and-Act with explicit replanning (LangGraph)
 * — TypeScript port of src/planning_plan_and_act.py.
 *
 * Anchor scenario: incident response automation.
 *
 * Flow:
 *   - planner node creates ordered actions
 *   - executor node performs one action at a time
 *   - reviewer node decides whether replanning is required
 */

import { ChatOpenAI } from "@langchain/openai";
import { END, StateGraph, Annotation } from "@langchain/langgraph";

const MODEL = process.env.OPENAI_MODEL ?? "gpt-4o-mini";

// ---------------------------------------------------------------------------
// State container for plan-and-act execution.
// ---------------------------------------------------------------------------

export const PlanActState = Annotation.Root({
  incident: Annotation<string>(),
  context: Annotation<string>(),
  plan: Annotation<string[]>({
    reducer: (_prev, next) => next,
    default: () => [],
  }),
  completed: Annotation<string[]>({
    reducer: (_prev, next) => next,
    default: () => [],
  }),
  current_action: Annotation<string>({
    reducer: (_prev, next) => next,
    default: () => "",
  }),
  notes: Annotation<string[]>({
    reducer: (_prev, next) => next,
    default: () => [],
  }),
  requires_replan: Annotation<boolean>({
    reducer: (_prev, next) => next,
    default: () => false,
  }),
});

type State = typeof PlanActState.State;

// ---------------------------------------------------------------------------
// LLM (lazy-init so smoke tests can import without OPENAI_API_KEY).
// ---------------------------------------------------------------------------

let _llm: ChatOpenAI | undefined;
function llm(): ChatOpenAI {
  if (!_llm) {
    _llm = new ChatOpenAI({ model: MODEL, temperature: 0 });
  }
  return _llm;
}

function contentToString(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((part) =>
        typeof part === "string"
          ? part
          : typeof part === "object" && part !== null && "text" in part
            ? String((part as { text: unknown }).text)
            : "",
      )
      .join("");
  }
  return String(content);
}

// ---------------------------------------------------------------------------
// Nodes
// ---------------------------------------------------------------------------

/** Create or refresh the execution plan. */
export async function plannerNode(state: State): Promise<Partial<State>> {
  const prompt = `
You are an incident commander. Build a short action plan (4-6 steps) for this incident.

Incident:
${state.incident}

Context:
${state.context}

Already completed:
${JSON.stringify(state.completed)}

Return one step per line, no numbering.
`;
  const response = await llm().invoke(prompt);
  const lines = contentToString(response.content).trim().split(/\r?\n/);
  const planRaw = lines
    .map((line) => line.replace(/^[-•\s]+/, "").trim())
    .filter((line) => line.length > 0);

  const plan = planRaw.filter((step) => !state.completed.includes(step));
  const notes = state.notes.concat(`planner: generated ${plan.length} remaining steps`);

  return { plan, notes, requires_replan: false };
}

/** Execute one action from the plan. */
export async function executorNode(state: State): Promise<Partial<State>> {
  if (state.plan.length === 0) {
    return { current_action: "" };
  }

  const action = state.plan[0]!;
  const remaining = state.plan.slice(1);

  const execPrompt = `
You are executing one incident response action.

Action: ${action}
Incident: ${state.incident}

Return exactly two lines:
STATUS: done|blocked
NOTE: <short operational note>
`;
  const result = await llm().invoke(execPrompt);
  const lines = contentToString(result.content)
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

  let status = "done";
  let note = "no note";
  for (const line of lines) {
    if (line.toLowerCase().startsWith("status:")) {
      status = line.slice(line.indexOf(":") + 1).trim().toLowerCase();
    }
    if (line.toLowerCase().startsWith("note:")) {
      note = line.slice(line.indexOf(":") + 1).trim();
    }
  }

  if (status === "done") {
    return {
      plan: remaining,
      current_action: action,
      completed: state.completed.concat(action),
      notes: state.notes.concat(`executor: done -> ${action} | ${note}`),
      requires_replan: false,
    };
  }
  return {
    plan: remaining,
    current_action: action,
    notes: state.notes.concat(`executor: blocked -> ${action} | ${note}`),
    requires_replan: true,
  };
}

/** Decide if another step can run or if replanning is needed. */
export function reviewerNode(state: State): Partial<State> {
  let note: string;
  if (state.requires_replan) {
    note = "reviewer: requesting replanning";
  } else if (state.plan.length > 0) {
    note = "reviewer: continue with remaining plan";
  } else {
    note = "reviewer: plan complete";
  }
  return { notes: state.notes.concat(note) };
}

export function routeAfterReview(state: State): "planner" | "executor" | typeof END {
  if (state.requires_replan) return "planner";
  if (state.plan.length > 0) return "executor";
  return END;
}

// ---------------------------------------------------------------------------
// Graph
// ---------------------------------------------------------------------------

export function buildGraph() {
  return new StateGraph(PlanActState)
    .addNode("planner", plannerNode)
    .addNode("executor", executorNode)
    .addNode("reviewer", reviewerNode)
    .addEdge("__start__", "planner")
    .addEdge("planner", "executor")
    .addEdge("executor", "reviewer")
    .addConditionalEdges("reviewer", routeAfterReview, {
      planner: "planner",
      executor: "executor",
      [END]: END,
    })
    .compile();
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  console.log("\n" + "=".repeat(80));
  console.log("PLANNING PATTERN — PLAN-AND-ACT (LANGGRAPH)");
  console.log("=".repeat(80));

  if (!process.env.OPENAI_API_KEY) {
    console.log("❌ OPENAI_API_KEY is required for this example.");
    return;
  }

  const initial = {
    incident: "Payment API error rate jumped to 35% after deployment.",
    context: "Service: billing-gateway, Region: eu-west, Recent change: release v2.8.4",
  };

  const app = buildGraph();
  const final = await app.invoke(initial);

  const completed = final.completed ?? [];
  const notes = final.notes ?? [];

  console.log("\n✅ Completed steps:");
  completed.forEach((step, idx) => {
    console.log(`  ${idx + 1}. ${step}`);
  });

  console.log("\n🧾 Execution notes:");
  for (const note of notes) {
    console.log(`  - ${note}`);
  }
}

if (import.meta.main) {
  await main();
}
